contracts/src/PDPService.sol

```diff
@@ -303,7 +302,29 @@ contract PDPService {
     // by the epoch of the previous proof of possession.
     // Note that this method is not restricted to the proof set owner.
     function provePossession(uint256 setId, Proof[] calldata proofs) public {
-        // TODO: implement me
+        uint256 challengeEpoch = nextChallengeEpoch[setId];
+        require(block.number >= challengeEpoch, "premature proof");
+
+        // TODO: fetch proper seed from chain randomness
+        uint256 seed = challengeEpoch;
+        uint256 leafCount = getProofSetLeafCount(setId);
+        uint256 sumTreeTop = 256 - BitOps.clz(nextRootId[setId]);
+
+        for (uint64 i = 0; i < proofs.length; i++) {
+            // Hash (SHA3) the seed,  proof set id, and proof index to create challenge.
+            bytes memory payload = abi.encodePacked(seed, setId, i);
+            uint256 challengeIdx = uint256(keccak256(payload)) % leafCount;
```

Comment:
This is an easy hash to use here, but we could use sha256 if preferred.

Reply:
sha3 lgtm